File: esm/engines/02_memory/cache.js
/**
 * SQLite-based response caching to save token costs.
 * Keyed by SHA-256(systemPrompt + prompt).
 * Upgraded to ESModules with seamless in-memory fallback.
 */

import fs from 'fs';
import path from 'path';
import crypto from 'crypto';

let DatabaseModule = null;
let db = null;
let useFallback = false;
let fallbackCache = new Map();
const fallbackFile = path.resolve(process.cwd(), 'data', 'cache.json');

try {
  // Try loading better-sqlite3 dynamically
  const betterSqlite = await import('better-sqlite3');
  DatabaseModule = betterSqlite.default;
  const dbPath = path.resolve(process.cwd(), 'data', 'cache.db');
  fs.mkdirSync(path.dirname(dbPath), { recursive: true });
  db = new DatabaseModule(dbPath);
  db.exec(`
    CREATE TABLE IF NOT EXISTS prompt_cache (
      key TEXT PRIMARY KEY,
      response TEXT NOT NULL,
      created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
    )
  `);
} catch (err) {
  console.warn(`[ResponseCache] SQLite database unavailable (${err.message}). Falling back to JSON cache file.`);
  useFallback = true;
  fs.mkdirSync(path.dirname(fallbackFile), { recursive: true });
  if (fs.existsSync(fallbackFile)) {
    try {
      const data = JSON.parse(fs.readFileSync(fallbackFile, 'utf8'));
      fallbackCache = new Map(Object.entries(data));
    } catch (_) {}
  }
}

export class ResponseCache {
  /**
   * Calculate SHA-256 key for prompts.
   * @param {string} systemPrompt 
   * @param {string} prompt 
   * @returns {string}
   */
  static calculateKey(systemPrompt, prompt) {
    const raw = `${systemPrompt}\x00${prompt}`;
    return crypto.createHash('sha256').update(raw).digest('hex');
  }

  /**
   * Fetch cached response, returns null if not found.
   * @param {string} systemPrompt 
   * @param {string} prompt 
   * @returns {string|null}
   */
  get(systemPrompt, prompt) {
    const key = ResponseCache.calculateKey(systemPrompt, prompt);
    if (useFallback) {
      return fallbackCache.get(key) || null;
    }

    try {
      const stmt = db.prepare('SELECT response FROM prompt_cache WHERE key = ?');
      const row = stmt.get(key);
      if (row) return row.response;
    } catch (e) {
      console.error(`[ResponseCache] Get cache failed: ${e.message}`);
    }
    return null;
  }

  /**
   * Cache a prompt response.
   * @param {string} systemPrompt 
   * @param {string} prompt 
   * @param {string} response 
   */
  set(systemPrompt, prompt, response) {
    const key = ResponseCache.calculateKey(systemPrompt, prompt);
    if (useFallback) {
      fallbackCache.set(key, response);
      try {
        const obj = Object.fromEntries(fallbackCache.entries());
        fs.writeFileSync(fallbackFile, JSON.stringify(obj, null, 2), 'utf8');
      } catch (e) {
        console.error(`[ResponseCache] Write fallback failed: ${e.message}`);
      }
      return;
    }

    try {
      const stmt = db.prepare('INSERT OR REPLACE INTO prompt_cache (key, response) VALUES (?, ?)');
      stmt.run(key, response);
    } catch (e) {
      console.error(`[ResponseCache] Set cache failed: ${e.message}`);
    }
  }
}

File: python/engines/memory/cache.py
"""
SQLite-based response caching to save token costs.
Keyed by SHA-256(system_prompt + prompt).
"""
import sqlite3
import os
import hashlib
import logging
from typing import Optional

logger = logging.getLogger(__name__)

class ResponseCache:
    """Manages SQLite cache for prompt responses."""
    
    def __init__(self, db_path: str = "data/cache.db"):
        self.db_path = db_path
        os.makedirs(os.path.dirname(self.db_path), exist_ok=True)
        self._init_db()

    def _init_db(self):
        with sqlite3.connect(self.db_path) as conn:
            conn.execute("""
                CREATE TABLE IF NOT EXISTS prompt_cache (
                    key TEXT PRIMARY KEY,
                    response TEXT NOT NULL,
                    created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
                )
            """)
            conn.commit()

    @staticmethod
    def calculate_key(system_prompt: str, prompt: str) -> str:
        """Calculate SHA-256 key for prompts."""
        raw = f"{system_prompt}\x00{prompt}".encode("utf-8")
        return hashlib.sha256(raw).hexdigest()

    def get(self, system_prompt: str, prompt: str) -> Optional[str]:
        """Fetch cached response, returns None if not found."""
        key = self.calculate_key(system_prompt, prompt)
        try:
            with sqlite3.connect(self.db_path) as conn:
                cursor = conn.cursor()
                cursor.execute("SELECT response FROM prompt_cache WHERE key = ?", (key,))
                row = cursor.fetchone()
                if row:
                    logger.debug("Cache hit for LLM response")
                    return row[0]
        except Exception as e:
            logger.error(f"Failed to read from cache: {e}")
        return None

    def set(self, system_prompt: str, prompt: str, response: str) -> None:
        """Cache a prompt response."""
        key = self.calculate_key(system_prompt, prompt)
        try:
            with sqlite3.connect(self.db_path) as conn:
                conn.execute(
                    "INSERT OR REPLACE INTO prompt_cache (key, response) VALUES (?, ?)",
                    (key, response)
                )
                conn.commit()
                logger.debug("LLM response cached successfully")
        except Exception as e:
            logger.error(f"Failed to write to cache: {e}")
